File: 23. Merge k Sorted Lists/solution.go
package main

func mergeKLists(lists []*ListNode) *ListNode {
	if len(lists) == 0 {
		return nil
	}

	for len(lists) > 1 {
		var tempMergedLists []*ListNode
		for i := 0; i < len(lists); i += 2 {
			l1 := lists[i]
			var l2 *ListNode
			if i+1 < len(lists) {
				l2 = lists[i+1]
			}

			l := merge2Lists(l1, l2)
			if l != nil {
				tempMergedLists = append(tempMergedLists, l)
			}
		}

		lists = tempMergedLists
	}

	if len(lists) == 0 {
		return nil
	} else {
		return lists[0]
	}
}

// return pointer can be nil
func merge2Lists(l1, l2 *ListNode) *ListNode {
	var first *ListNode

	if l1 == nil {
		return l2
	} else if l2 == nil {
		return l1
	} else if l1.Val <= l2.Val {
		first = l1
		l1 = l1.Next
	} else {
		first = l2
		l2 = l2.Next
	}

	head := first
	for l1 != nil && l2 != nil {
		if l1.Val <= l2.Val {
			head.Next = l1
			l1 = l1.Next
		} else {
			head.Next = l2
			l2 = l2.Next
		}
		head = head.Next
		head.Next = nil
	}

	if l1 != nil {
		head.Next = l1
	} else if l2 != nil {
		head.Next = l2
	}

	return first
}

File: 23. Merge k Sorted Lists/solution.js
function merge2Lists(l1, l2) {
    let start;
    if (l1 === undefined || l1 === null) {
        return l2;
    } else if (l2 === undefined || l2 === null) {
        return l1;
    } else if (l1.val <= l2.val) {
        start = l1;
        l1 = l1.next;
    } else {
        start = l2;
        l2 = l2.next;
    }

    let head = start;
    while (l1 !== undefined && l1 !== null && l2 !== undefined && l2 !== null) {
        if (l1.val <= l2.val) {
            head.next = l1
            l1 = l1.next;
        } else {
            head.next = l2;
            l2 = l2.next;
        }
        head = head.next;
        head.next = null;
    }

    if (l1 !== undefined && l1 !== null) {
        head.next = l1;
    } else if (l2 !== undefined && l2 !== null) {
        head.next = l2;
    }

    return start;
}

var mergeKLists = function(lists) {
    if (lists === undefined || lists.length === 0) {
        return null;
    }

    while (lists.length > 1) {
        let temp = [];
        for (i = 0; i < lists.length; i += 2) {
            let l1 = lists[i]
            let l2;
            if (i + 1 < lists.length) {
                l2 = lists[i + 1];
            }
            let merged = merge2Lists(l1, l2);
            if (merged !== undefined && merged !== null) {
                temp.push(merged);
            }
        }
        lists = temp;
    }

    if (lists.length === 0) {
        return null;
    } else {
        return lists[0]
    }
}